### PR TITLE
[#7] As a user, I can run unit test

### LINKED ISSUE #7: As a user, I can run unit test
## Why
- Ensure all the basic requirement for unit testing is set up (mockito, test_api, test directory).
- Create a basic test case as a showcase to the developer who will use this template.
 
## Who Benefits?
Every developer who uses this template.

### PR DESCRIPTION
**Pull request by manh-t**
Close #7

## What happened 👀

Ensure all the basic requirement for unit testing is set up (mockito, test_api, test directory).
Create a basic test case as a showcase to the developer who will use this template.
 
## Insight 📝

I create temporary `api_service` and `credential_repository` then perform unit testing on `credential_repository`.
These files might be updated to reflect the real implementation.
 
## Proof Of Work 📹

Passed 2/2
image

### CODE REVIEW

File: test/repository/credential_repository_test.dart [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    test("When getting profile failed, it throws a negative error", () async {
+      when(mockApiService.getProfile()).thenAnswer(
+          (_) => Future.error(TimeoutException('There is an exception!')));
+
+      try {
+        await repository.getProfile();
+      } catch (e) {
+        expect(e, isInstanceOf<TimeoutException>());
+        expect((e as TimeoutException).message, 'There is an exception!');
```
Comment (sleepylee) [inline test/repository/credential_repository_test.dart:37]: maybe do this:

```
expect(() => repository.getProfile(), throwsA<TimeoutException>(predicate((ex) => ex.message == 'There is an exception!')));
```

shorter? 💁‍♂️

Comment (manh-t) [reply test/repository/credential_repository_test.dart:37]: ehh yes, it's cleaner 👍

Comment (sleepylee) [reply test/repository/credential_repository_test.dart:37]: you can do a short declaration as well:

```
final result = () => repository.getProfile();
expect(result, ....)

Comment (manh-t) [reply test/repository/credential_repository_test.dart]: Updated 🙏

File: test/repository/credential_repository_test.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      when(mockApiService.getProfile())
+          .thenAnswer((_) async => UserResponse('test@email.com', 'test_user'));
+
+      final result = await repository.getProfile();
+      expect(result.email, 'test@email.com');
+      expect(result.username, 'test_user');
+    });
+
+    test("When getting profile failed, it throws a negative error", () async {
+      when(mockApiService.getProfile()).thenAnswer(
```
Comment (sleepylee) [inline test/repository/credential_repository_test.dart:30]: `thenThrow(TimeoutException('There is an exception!'));` - I forgot this as well 😛

File: test/repository/credential_repository_test.dart [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import 'package:flutter_templates/model/response/user_response.dart';
+import 'package:flutter_templates/repository/credential_repository.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:mockito/mockito.dart';
+
+import '../mocks/generate_mocks.mocks.dart';
+
+void main() {
+  group('CredentialRepository', () {
```
Comment (luongvo) [inline test/repository/credential_repository_test.dart:11]: should this group name is `Test getProfile`? Then we will have another group is `Test updateProfile` for example 🤓

**Review by luongvo — COMMENTED** [review]
rest lgtm

File: test/repository/credential_repository_test.dart (hunk 3 above)
Comment (manh-t) [reply test/repository/credential_repository_test.dart:11]: @luongvo so for each function, we will test it in a group 💭 if so, I think we can go with `CredentialRepository - getProfile`. It's clearer and we are doing test so we don't need `Test` word, how do you think?

Comment (luongvo) [reply test/repository/credential_repository_test.dart:11]: @manh-t we don't need to repeat `CredentialRepository` as we're on this CredentialRepositoryTest already. Does that `CredentialRepository` prefix in the group name help on anything? 🤔

Comment (manh-t) [reply test/repository/credential_repository_test.dart:11]: @luongvo for example, when we run command `flutter test` and there is a failed test case. It's hard to see which test part is failed, unless we define them in the group name
image

Comment (luongvo) [reply test/repository/credential_repository_test.dart:11]: @manh-t so it's okay to keep `CredentialRepository` or group them as `CredentialRepository - getProfile`

**Comment by manh-t** [comment]
Updated 👔

**Review by luongvo — APPROVED** [review]
lgtm

**Review by sleepylee — APPROVED** [review]

**Pull request merged**